### DISCUSSION TITLE
Reconstruct dual infeasibility from results

### DISCUSSION THREAD

**Discussion by XingchenFan** [coin-or/Ipopt#608] [Q&A] [answered]
Hi all,

I am using the compilation of IPOPT from [OPTI](https://github.com/jonathancurrie/OPTI) to solve a large NLP in MATLAB. Very often, the dual infeasibility doesn't converge hence failing the solve. I want to manually calculate the dual infeasibility from results so I can find which variables and constraints cause this. Unfortunately, my naive attempt doesn't generate the same dual infeasibility as reported from the command line print-out.

Here is what I did and please let me know what I should fix:

The ipopt function returns:
1. Results: `x`
2. A summary structure with: `zl` and `zu` (multipliers for lower and upper bounds), `lambda` (multipliers for equality constraints)

I use x to evaluate objective gradient `df` and constraint Jacobian `dg` (would be nice if it can be returned directly...)

Then use formula from the IPOPT reference paper (On the Implementation of an Interior-Point Filter Line-Search Algorithm for Large-Scale Nonlinear Programming) Eqn 4a:

> inf_du = max(abs(df + dg' * lambda - zl - zu))

As I mentioned earlier, this doesn't give the same _inf_du_ from the print-out. One thing I know I am missing is the scaling, since the print-out value is scaled. But setting _nlp_scaling_method_ to 'none' doesn't help.

So am I using the right approach? What am I missing?

**Comment by svigerske** [comment, marked as answer]
From the code, it is `+ zu` and not `- zu`.

This is the function that calculates the `inf_du` from the output (calles with `NormType=NormMax`):
https://github.com/coin-or/Ipopt/blob/stable/3.14/src/Algorithm/IpIpoptCalculatedQuantities.cpp#L2653-L2697

The calculation of the gradient of the Lagrangian is at https://github.com/coin-or/Ipopt/blob/stable/3.14/src/Algorithm/IpIpoptCalculatedQuantities.cpp#L1993-L2030, in particular
```cpp
         tmp->Copy(*curr_grad_f());
         tmp->AddTwoVectors(1., *curr_jac_cT_times_curr_y_c(), 1., *curr_jac_dT_times_curr_y_d(), 1.);
         DBG_PRINT_VECTOR(2, "jac_cT*y_c", *curr_jac_cT_times_curr_y_c());
         DBG_PRINT_VECTOR(2, "jac_dT*y_d", *curr_jac_dT_times_curr_y_d());
         ip_nlp_->Px_L()->MultVector(-1., *z_L, 1., *tmp);
         ip_nlp_->Px_U()->MultVector(1., *z_U, 1., *tmp);
```
So, in some sense, it is `df + dg' lambda - zl + zu`.

Note that Ipopt also works on an [internal reformulation of the problem](https://coin-or.github.io/Ipopt/classIpopt_1_1TNLPAdapter.html#details) with slack variables added. This adds another term to the dual infeasibility, using quantities that are probably not accessible in OPTI (vl, vu).
